core/src/main/java/io/temporal/samples/keymanagementencryption/awsencryptionsdk/KeyringCodec.java

```diff
@@ -0,0 +1,155 @@
+   *
+   * @param kmsKeyring the keyring to use for encryption and decryption.
+   */
+  public KeyringCodec(IKeyring kmsKeyring) {
+    this.crypto = AwsCrypto.standard();
+    this.kmsKeyring = kmsKeyring;
+    this.useSerializationContext = false;
+    this.serializationContext = null;
+  }
+
+  /**
+   * Constructs a new KeyringCodec with the provided {@link IKeyring}.
+   *
+   * @param crypto the AWS Crypto object to use for encryption and decryption.
+   * @param kmsKeyring the keyring to use for encryption and decryption.
+   * @param useSerializationContext whether to use a {@link SerializationContext} for encoding and
+   *     decoding payloads.
+   */
+  public KeyringCodec(AwsCrypto crypto, IKeyring kmsKeyring, boolean useSerializationContext) {
+    this.crypto = crypto;
+    this.kmsKeyring = kmsKeyring;
+    this.useSerializationContext = useSerializationContext;
+    this.serializationContext = null;
+  }
+
+  private KeyringCodec(
+      AwsCrypto crypto, IKeyring kmsKeyring, SerializationContext serializationContext) {
+    this.crypto = crypto;
+    this.kmsKeyring = kmsKeyring;
+    this.useSerializationContext = true;
+    this.serializationContext = serializationContext;
+  }
+
+  @NotNull
+  @Override
+  public List<Payload> encode(@NotNull List<Payload> payloads) {
+    return payloads.stream().map(this::encodePayload).collect(Collectors.toList());
+  }
+
+  @NotNull
+  @Override
+  public List<Payload> decode(@NotNull List<Payload> payloads) {
+    return payloads.stream().map(this::decodePayload).collect(Collectors.toList());
+  }
+
+  @NotNull
+  @Override
+  public PayloadCodec withContext(@Nonnull SerializationContext context) {
+    if (!useSerializationContext) {
+      return this;
+    }
+    return new KeyringCodec(crypto, kmsKeyring, context);
+  }
+
+  private Map<String, String> getEncryptionContext() {
+    // If we are not using a serialization context, return an empty map
+    // There may not be a serialization context if certain cases, such as when the codec is used
+    // for encoding/decoding payloads for a Nexus operation.
+    if (!useSerializationContext
+        || serializationContext == null
+        || !(serializationContext instanceof HasWorkflowSerializationContext)) {
+      return Collections.emptyMap();
+    }
+    String workflowId = ((HasWorkflowSerializationContext) serializationContext).getWorkflowId();
+    String activityType = null;
+    if (serializationContext instanceof ActivitySerializationContext) {
+      activityType = ((ActivitySerializationContext) serializationContext).getActivityType();
+    }
+    String signature = activityType != null ? workflowId + activityType : workflowId;
+    return Collections.singletonMap("signature", signature);
+  }
+
+  private Payload encodePayload(Payload payload) {
+    byte[] plaintext = payload.toByteArray();
+    byte[] ciphertext =
+        crypto.encryptData(kmsKeyring, plaintext, getEncryptionContext()).getResult();
```

Comment:
Is this call potentially slow (i.e. meaning ever makes a network call, even once lazily)? Should the deadlock detector be disabled or is it already disabled?

Reply:
Yeah good point, I added a line to disable the deadlock detector.